lib/block-supports/variations.php

```diff
@@ -0,0 +1,341 @@
+
+	return gutenberg_merge_block_style_variations_data( $variations_data, $theme_json );
+}
+
+/**
+ * Merges shared block style variations registered within the
+ * `styles.blocks.variations` property of the primary theme.json file.
+ *
+ * @param WP_Theme_JSON_Data_Gutenberg $theme_json Current theme.json data.
+ *
+ * @return WP_Theme_JSON_Data_Gutenberg
+ */
+function gutenberg_resolve_block_style_variations_from_primary_theme_json( $theme_json ) {
+	$theme_json_data        = $theme_json->get_data();
+	$block_style_variations = $theme_json_data['styles']['blocks']['variations'] ?? array();
+	$variations_data        = gutenberg_resolve_and_register_block_style_variations( $block_style_variations );
+
+	return gutenberg_merge_block_style_variations_data( $variations_data, $theme_json );
+}
+
+/**
+ * Merges block style variations registered via the block styles registry with a
+ * style object, under their appropriate block types within theme.json styles.
+ * Any variation values defined within the theme.json specific to a block type
+ * will take precedence over these shared definitions.
+ *
+ * @param WP_Theme_JSON_Data_Gutenberg $theme_json Current theme.json data.
+ *
+ * @return WP_Theme_JSON_Data_Gutenberg
+ */
+function gutenberg_resolve_block_style_variations_from_styles_registry( $theme_json ) {
+	$registry        = WP_Block_Styles_Registry::get_instance();
+	$styles          = $registry->get_all_registered();
+	$variations_data = array();
+
+	foreach ( $styles as $block_type => $variations ) {
+		foreach ( $variations as $variation_name => $variation ) {
+			if ( ! empty( $variation['style_data'] ) ) {
+				$path = array( $block_type, 'variations', $variation_name );
+				_wp_array_set( $variations_data, $path, $variation['style_data'] );
+			}
+		}
+	}
+
+	return gutenberg_merge_block_style_variations_data( $variations_data, $theme_json );
+}
+
+/**
+ * Enqueues styles for block style variations.
+ */
+function gutenberg_enqueue_variation_styles() {
+	wp_enqueue_style( 'variation-styles' );
+}
+
+// Register the block support.
+WP_Block_Supports::get_instance()->register( 'variation', array() );
+
+add_filter( 'render_block_data', 'gutenberg_render_variation_support_styles', 10, 2 );
+add_filter( 'render_block', 'gutenberg_render_variation_class_name', 10, 2 );
+add_action( 'wp_enqueue_scripts', 'gutenberg_enqueue_variation_styles', 1 );
+
+// Resolve block style variations from all their potential sources. The order here is deliberate.
+add_filter( 'wp_theme_json_data_theme', 'gutenberg_resolve_block_style_variations_from_primary_theme_json', 10, 1 );
+add_filter( 'wp_theme_json_data_theme', 'gutenberg_resolve_block_style_variations_from_theme_json_partials', 10, 1 );
+add_filter( 'wp_theme_json_data_theme', 'gutenberg_resolve_block_style_variations_from_styles_registry', 10, 1 );
+
+add_filter( 'wp_theme_json_data_user', 'gutenberg_resolve_block_style_variations_from_theme_style_variation', 10, 1 );
```

Comment:
optional: I think my preference is to locate these next to the function declarations so that it's easy to tell which filter the function is for when reading the function body.

Reply:
I can see the argument for that. Given the importance of filter order, I'd rather the filters together in this case. Locating the filter by the functions isn't exactly followed for all the other block supports either.